### PR TITLE
[InstCombine] Fold select with signbit idiom into fabs

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch folds:
```
((bitcast X to int) <s 0 ? -X : X) -> fabs(X)
((bitcast X to int) >s -1 ? X : -X) -> fabs(X)
((bitcast X to int) <s 0 ? X : -X) -> -fabs(X)
((bitcast X to int) >s -1 ? -X : X) -> -fabs(X)
```
Alive2: https://alive2.llvm.org/ce/z/rGepow

### CODE REVIEW

**Comment by dtcxzyw** [comment]
Ping.

**Comment by dtcxzyw** [comment]
Ping.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 1]
```diff
@@ -2763,6 +2763,41 @@ static Instruction *foldSelectWithFCmpToFabs(SelectInst &SI,
...
+      return nullptr;
+
+    // Fold (IsSigned ? -X : X) or (!IsSigned ? X : -X) to fabs(X)
+    // Fold (IsSigned ? X : -X) or (!IsSigned ? -X : X) to -fabs(X)
+    Value *Fabs = IC.Builder.CreateUnaryIntrinsic(Intrinsic::fabs, X, &SI);
+    if (Swap != TrueIfSigned)
+      return IC.replaceInstUsesWith(SI, Fabs);
+    else {
+      Instruction *NewFNeg = UnaryOperator::CreateFNeg(Fabs);
+      NewFNeg->setFastMathFlags(SI.getFastMathFlags());
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:2796]: CreateFNegFMF?

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 2]
```diff
@@ -2763,6 +2763,41 @@ static Instruction *foldSelectWithFCmpToFabs(SelectInst &SI,
     }
   }
 
+  // Match select with (icmp slt (bitcast X to int), 0)
+  //                or (icmp sgt (bitcast X to int), -1)
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:2767]: I suppose it would also work for `fptosi`?

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:2767]: Yeah. But this pattern is usually generated by `std::signbit`. Not sure that `((fptosi X) <s 0 ? -X : X)` exists in real-world applications.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 3]
```diff
@@ -2763,6 +2763,41 @@ static Instruction *foldSelectWithFCmpToFabs(SelectInst &SI,
...
+      return nullptr;
+    if (!CondVal->hasOneUse() && !TrueVal->hasOneUse())
+      return nullptr;
+
+    // Fold (IsSigned ? -X : X) or (!IsSigned ? X : -X) to fabs(X)
+    // Fold (IsSigned ? X : -X) or (!IsSigned ? -X : X) to -fabs(X)
+    Value *Fabs = IC.Builder.CreateUnaryIntrinsic(Intrinsic::fabs, X, &SI);
+    if (Swap != TrueIfSigned)
+      return IC.replaceInstUsesWith(SI, Fabs);
+    else {
```
[the remark this answers is not in the document]
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp]: Done.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 4]
```diff
@@ -2763,6 +2763,41 @@ static Instruction *foldSelectWithFCmpToFabs(SelectInst &SI,
...
+    Value *X = SI.getFalseValue();
+
+    if (Swap)
+      std::swap(TrueVal, X);
+
+    CmpInst::Predicate Pred;
+    const APInt *C;
+    bool TrueIfSigned;
+    if (!match(CondVal, m_ICmp(Pred, m_BitCast(m_Specific(X)), m_APInt(C))) ||
+        !IC.isSignBitCheck(Pred, *C, TrueIfSigned))
```
[the remark this answers is not in the document]
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:2786]: Done.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp (hunk 1 above)
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp]: Done.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 5]
```diff
@@ -2763,6 +2763,41 @@ static Instruction *foldSelectWithFCmpToFabs(SelectInst &SI,
...
+      continue;
+    if (!match(TrueVal, m_FNeg(m_Specific(X))))
+      return nullptr;
+    if (!CondVal->hasOneUse() && !TrueVal->hasOneUse())
```
[the remark this answers is not in the document]
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp]: I believe it is always a win if we don't create more instructions than we save.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:2786]: I am sorry I cannot get your point. Can you explain further?

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp (hunk 2 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:2771]: I agree the bitcast one is probably alot more likely, but could imagine some template code casting float -> int for a compare.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:2786]: Ping.

**Review by arsenm — APPROVED** [review]

**Pull request merged**